fix: Add better version detection

ngirardin opened the pull request:
This is a follow-up PR for #118 which have broken some installs:
- https://github.com/vitest-dev/vscode/issues/88#issuecomment-1398074211

It seems to fix:
- https://github.com/vitest-dev/vscode/issues/124
- https://github.com/vitest-dev/vscode/issues/124
- https://github.com/vitest-dev/vscode/issues/90
- https://github.com/vitest-dev/vscode/issues/131

and probably:
- https://github.com/vitest-dev/vscode/issues/13

In some cases, vitest need to be started using `spawn(vitestCommand.cmd, [...vitestCommand.args, '-v'], envs))`, and in other by using `spawn(process.execPath, [vitestCommand.cmd, ...vitestCommand.args, '-v'], envs))`.

There are lot of OS / Node install combinations, and I don't have a clear picture of which to use and when.

The easiest fix that I could find, was to try both commands and see which one worked.

It's not the cleanest solution, but I'm open to suggestions :)

ngirardin commented:
@zxch3n any thoughts on the fix? (I can't add you as a reviewer)

zxch3n commented on src/pure/utils.ts:
```diff
@@ -80,33 +80,48 @@ export interface Cmd {
   }
+}
+
+const getVersionOrThrow = (vitestCommand: string | undefined): string => {
```
The arg does not seem like vitestCommand

zxch3n commented on src/pure/utils.ts:
```diff
@@ -80,33 +80,48 @@ export interface Cmd {
+
+  const version = getVersionOrThrow(await spawnVitestVersion(vitestCommand.cmd, [...vitestCommand.args, '-v'], envs))
+
+  if (version !== undefined)
```
It's impossible to be undefined here

zxch3n reviewed (changes requested):
This fix should use `try` and `catch` to avoid throwing in each step, right?

arimgibson commented:
Hey @ngirardin , let me know if you need help making those suggestions from @zxch3n ; happy to help move this forward if there's anything I can do!

ngirardin commented:
Thanks @zxch3n for the review.

I've made some changes to the code:
- log the commands that are spawned
- added a message to invite users to open an issue and provide their logs if no version is detected
- trying both of the spawn methods for starting `npx` (when `command == null`)
- added some tests (WIP needs to mock spawn to test all the cases)
- added a check when doing the `line.match` to avoid crashing if the version is not found in the output

@arimgibson, sorry for the delay, please let me know if you see something to change on this code :)

zxch3n commented on src/pure/utils.ts:
```diff
@@ -80,31 +81,73 @@ export interface Cmd {
+  })
 
   // eslint-disable-next-line no-unreachable-loop
   for await (const line of chunksToLinesAsync(child.stdout)) {
```
Maybe you should use try catch to wrap 106~114 here? Otherwise, it may just throw on 127 or 133 line

ngirardin replied:
Is `chunksToLinesAsync` throwing if the `spawn` fails?

When testing locally, if the spawn fails, we're not entering in the `for` loop and `detectVitestVersion` run the second command.

zxch3n replied:
I was afraid of the for await might throw. If you've tested it, that's fine. Thanks!

ngirardin replied:
Thank you for the review and the merge.

I'll add some tests, so we'll be able to test the edge cases if some new bugs are open, I found a lib to mock `spawn`
